webrtc/README.md

```diff
@@ -0,0 +1,199 @@
+
+4. _A_ establishes the connection to _B_. Yet to be determined how _B_ learns of
+   _A_ fingerprint. Potentially through A's STUN message USERNAME attribute. See
+   Open Questions below.
+
+5. _A_ initiates some authentication handshake _X_ to _B_ on a datachannel,
+   where _X_ allows _A_ and _B_ to authenticate each other's peer IDs. _X_ could
+   for example be Noise. See WebTransport specification as an example
+   https://github.com/libp2p/specs/pull/404. Still to be specified here for
+   WebRTC.
+
+6. On success of the authentication handshake _X_, the used datachannel is
+   closed and the plain WebRTC connection is used with its multiplexing
+   capabilities via datachannels.
+
+### Browser to Browser
+
+Scenario: Browser _A_ wants to connect to Browser node _B_ with the help of
+server node _R_.
+
+- Replace STUN with libp2p's identify and AutoNAT
+  - https://github.com/libp2p/specs/tree/master/identify
+  - https://github.com/libp2p/specs/blob/master/autonat/README.md
+- Replace TURN with libp2p's Circuit Relay v2
+  - https://github.com/libp2p/specs/blob/master/relay/circuit-v2.md
+
+## Open Questions
+
+_Where _Browser_ is replaced with the major desktop browsers of today (Chrome,
+Safari, Edge, Firefox)._
+
+- Should libp2p's WebRTC stack limit itself to using UDP only, or support WebRTC
+  on top of both UDP and TCP?
+
+### Direct connections
+
+#### Browser to public server node
+
+- Can _Browser_ generate a _valid_ SDP packet for the remote node based on the
+  remote's Multiaddr, where that Multiaddr contains the IP, UDP port and TLS
+  certificate fingerprint (e.g.
+  `/ip6/2001:db8::/udp/1234/webrtc/certhash/<hash>/p2p/<peer-id>`)? _Valid_ in
+  the sense that this generated SDP packet can then be used to establish a
+  WebRTC connection to the remote.
+
+- Do the major (Go / Rust / ...) WebRTC implementations allow us to accept a
+  WebRTC connection from a remote node without previously receiving an SDP
+  packet from such host?
+
+  How does the server learn the TLS certificate fingerprint of the browser? Is
+  embedding A's TLS certificate fingerprint in A's STUN message USERNAME
+  attribute a valid option?
+
+#### Browser to browser or non-public server node
+
+- Can _Browser_ know upfront its UDP port which it is listening for incoming
+  connections on? Does the browser reuse the UDP port across many WebRTC
+  connections? If that is the case one could connect to any public node, with
+  the remote telling the local node what port it is perceived on.
+
+- Can _Browser_ control the lifecycle of its local TLS certificate, i.e. can
+  _Browser_ use the same TLS certificate for multiple WebRTC connections?
+
+- Can two _Browsers_ exchange their SDP packets via a third server node using
+  Circuit Relay v2 and DCUtR? Instead of exchanging the original SDP packets,
+  could they exchange their multiaddr and construct the remote's SDP packet
```

Comment:
consider two browsers _A_ and _B_ . If _A_ wishes to connect to _B_, it would have to generate SDP (or at the very least generate credentials, but this cannot be done without generating SDP) and send it to _B_. _B_ would need to instantiate a new `RTCPeerConnection` object to process incoming data from _A_.

Reply:
> Apart from this, they would have to exchange server-reflexive candidates which are not present in the SDP.

Can you expand on this for a WebRTC beginner? I would expect, once a browser learned its server-reflexive addresses from some STUN server, that the locally generated SDP would contain those learned addresses. Is that not the case @ckousik?

Reply:
> I would expect, once a browser learned its server-reflexive addresses from some STUN server, that the locally generated SDP would contain those learned addresses.

This is my current thinking as well.

Reply:
Thanks for expanding here @ckousik! The above makes sense to me.

Reply:
yes, but ideally you should wait for a `srvflx` or `relayed` candidate if you are using a STUN or TURN server. The first callback could just contain the host candidate which may not be externally accessible.

Reply:
Resolving here with the recent updates to the specification. Please start a new discussion in case I am missing something.